doc/design/session.md

```diff
@@ -0,0 +1,62 @@
+# Design Doc: Session
+
+## Abstract
+
+This design doc proposes to have an object called *Session* which
+encapsulates the environment in which the computation graph is
+executed.
+
+## Background
+
+A computation graph is executed in an environment which contains the
+[scope](./scope.md) and other states. PaddlePaddle used to only have
+an implicit global session on which `paddle.eval()` is executed.
+
+This has the limitation that the user can not create two independent
+environments. For example, in reinforcement learning, the user may
+want to have a stale model for inference and a fresh model for
+training, and only replace the stale model with the fresh model
+periodically. Also, we have no concept that can encapsulate a remote
+environment that could execute a computation graph.
+
+## Session
+
+Session is an object that owns all runtime states such as scope,
+reader OP's file handles, connection to a remote PaddlePaddle cluster,
+etc.
+
+Session has two methods: `eval` and `close`. `eval` executes the
+target OP in a given graph, and `close` closes the session and
+releases all related resources:
+
+```Python
+a = paddle.constant(1.0)
+b = paddle.constant(2.0)
+c = a + b
+sess = paddle.session()
```

Comment:
Some thoughts of mine:

- An NN training job must contain: graph(containing sub nets like init_net, forward_net, backward_net, opt_net); scope containing tensors as parameters; hyper parameters(learning_rate, batch_size, etc.); settings(cluster or not, devices, quotas, node ip address etc.), so here may be something like:
      sess = paddle.session(graph, scope_list, settings)
      # or
      sess = paddle.remote_session(graph, scop_list, cluster_settings)

- states can be stored all in some "scope" by creating scope for storing tensor for forward and backward,  and for storing hyper parameters changing states.

Reply:
Agree with
```
sess = paddle.session(graph, scope_list, settings)
# or
sess = paddle.remote_session(graph, scop_list, cluster_settings)
```

> states can be stored all in some "scope" by creating scope for storing tensor for forward and backward, and for storing hyper parameters changing states.

Yes the variable states are stored in "scope". One session means one scope (just added into the doc).